src/coreclr/jit/lowerxarch.cpp

```diff
@@ -9528,22 +9545,44 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                             // result = ([op1] * op2) + op3
                             MakeSrcContained(node, op1);
                         }
-                        else if (supportsOp3RegOptional)
+                        else if (supportsOp1RegOptional)
                         {
-                            assert(resultOpNum != 3);
-                            MakeSrcRegOptional(node, op3);
+                            if (supportsOp2RegOptional)
+                            {
+                                GenTree* regOptionalOperand = PreferredRegOptionalOperand(op1, op2);
 
-                            // TODO-XArch-CQ: Any operand can be reg-optional.
-                            //                https://github.com/dotnet/runtime/issues/6358
+                                if (supportsOp3RegOptional)
```

Comment:
can we make this logic little simpler by capturing the `regOptionalOperand` appropriately and then just use `MakeSrcRegOptional(node, regOptionalOperand);` once?

Reply:
It's not much simpler to do so and actually results in longer code, which is why I opted to not do it.

Ultimately we have 3 conditions to consider and we need to ensure we only consider a given operand if it supports being marked `regOptional`, this means the alternative looks like:
```cpp
else
{
    GenTree* regOptionalOperand = nullptr;

    if (supportsOp1RegOptional)
    {
        if (supportsOp2RegOptional)
        {
            regOptionalOperand = PreferredRegOptionalOperand(op1, op2);

            if (supportsOp3RegOptional)
            {
                regOptionalOperand = PreferredRegOptionalOperand(regOptionalOperand, op3);
            }
        }
        else if (supportsOp3RegOptional)
        {
            regOptionalOperand = PreferredRegOptionalOperand(op1, op3);
        }
        else
        {
            regOptionalOperand = op1;
        }
    }
    else if (supportsOp2RegOptional)
    {
        if (supportsOp3RegOptional)
        {
            regOptionalOperand = PreferredRegOptionalOperand(op2, op3);
        }
        else
        {
            regOptionalOperand = op2;
        }
    }
    else if (supportsOp3RegOptional)
    {
        regOptionalOperand = op3;
    }

    if (regOptionalOperand != nullptr)
    {
        MakeSrcRegOptional(node, regOptionalOperand);
    }
}
```

So we removed 3x `MakeSrcRegOptional` calls and replaced 3x other `MakeSrcRegOptional` calls with local assignments. This comes at the cost of increasing nesting and introducing an additional required branch into the code path. It doesn't actually save in terms of the number of calls being done, that still remains just 1x call to `MakeSrcRegOptional` regardless of pattern.